crates/subspace-service/src/domains/snap_sync_orchestrator.rs

```diff
@@ -0,0 +1,122 @@
+//! Provides synchronization primitives for consensus and domain chains snap sync.
+
+use parking_lot::Mutex;
+use subspace_core_primitives::BlockNumber;
+use tokio::sync::Notify;
+use tracing::debug;
+
+/// Synchronizes consensus and domain chain snap sync.
+pub struct SnapSyncOrchestrator {
+    notify_consensus_snap_sync_unblocked: Notify,
+    consensus_snap_sync_block_number: Mutex<Option<BlockNumber>>,
+    notify_domain_snap_sync_unblocked: Notify,
+    notify_other_consensus_sync_strategies_unblocked: Notify,
+    notify_domain_snap_sync_finished: Notify,
+    domain_snap_sync_finished: Mutex<bool>,
+}
+
+impl Default for SnapSyncOrchestrator {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl SnapSyncOrchestrator {
+    /// Constructor
+    pub fn new() -> Self {
+        Self {
+            notify_consensus_snap_sync_unblocked: Notify::new(),
+            consensus_snap_sync_block_number: Mutex::new(None),
+            notify_domain_snap_sync_unblocked: Notify::new(),
+            notify_other_consensus_sync_strategies_unblocked: Notify::new(),
+            notify_domain_snap_sync_finished: Notify::new(),
+            domain_snap_sync_finished: Mutex::new(false),
+        }
+    }
+
+    /// Returns optional target block for consensus chain snap sync. None means target block is
+    /// not defined yet.
+    pub fn target_consensus_snap_sync_block_number(&self) -> Option<BlockNumber> {
+        *self.consensus_snap_sync_block_number.lock()
+    }
+
+    /// Wait for the allowing signal for the consensus chain snap sync.
+    pub async fn consensus_snap_sync_unblocked(&self) {
```

Comment:
`SubspaceLink::block_importing_notification_stream()`

Reply:
Why not deleting this one then? My point was that we ideally wouldn't need this orchestrator at all.

Reply:
Consensus chain snap sync is a part of the more complex domain snap sync process. In this form, it must start after we acquire the correct target block. I removed other blocking orchestrator points after our conversation: for example, we don't need to send signals from consensus snap sync anymore, however, it's not clear to me how to remove the dependency completely. Let's wait until the full solution is merged and return to this, I'm open for a change here.

Reply:
Can you just have a mutex or oneshot channel or something that is passed down to `subspace-service` that prevents sync as such from starting? I don't think you need to block/unblock it many times anyway, just pause until something happens on domain side and it is not necessarily specific to Snap sync either.

What else is this orchestrator needed for?

Reply:
The current implementation will contain the target block provider that conceals the orchestrator in the full version:
```
pub trait SnapSyncTargetBlockProvider: Send + Sync {
    async fn target_block(&self) -> Option<BlockNumber>;
}
```
The default non-blocking implementation returns `None,` which is close to what you proposed. I tried to limit the scope of the PR, but it seems the whole solution will provide more context and will be easier to review despite its size.